src/dune/file_tree.ml

```diff
@@ -538,16 +650,34 @@ module Dir = struct
     (Memo.Cell.get_sync s.sub_dir_as_t |> Option.value_exn).dir
 
   let fold_sub_dirs (t : t) ~init ~f =
-    String.Map.foldi t.contents.sub_dirs ~init ~f:(fun name s acc ->
-        f name (sub_dir_as_t s) acc)
+    String.Map.foldi t.contents.sub_dirs ~init ~f:(fun basename s acc ->
+        f ~basename (sub_dir_as_t s) acc)
+
+  let fold_dune_files (type acc) t ~(init : acc) ~f =
+    let rec loop ~basename dir (acc : acc) : acc =
+      let init =
+        match dune_file dir with
+        | None -> acc
+        | Some dune_file -> f ~basename dir dune_file acc
+      in
+      fold_sub_dirs dir ~init ~f:(fun ~basename ->
+          loop ~basename:(Some basename))
+    in
+    let basename =
+      if Path.Source.is_root t.path then
```

Comment:
I think we should make a function `Path.Source.basename_opt` for this, btw. (or, in fact, call it `basename` and rename the old `basename` to `basename_exn`.)

Reply:
I've introduced `Path.Source.basename_opt`. Let's keep the renaming to another PR.